plugins/regex_revalidate/regex_revalidate.c

```diff
@@ -221,6 +229,96 @@ prune_config(invalidate_t **i)
   return pruned;
 }
 
+static bool
+load_state(plugin_state_t *pstate, invalidate_t **ilist)
+{
+  FILE *fs = NULL;
+  struct stat s;
+  char line[LINE_MAX];
+  time_t now;
+  pcre *config_re = NULL;
+  const char *errptr;
+  int erroffset, ovector[OVECTOR_SIZE], rc;
+  int ln = 0;
+
+  if (!*ilist) {
+    return true;
+  }
+
+  if (stat(pstate->state_path, &s) < 0) {
+    TSDebug(PLUGIN_NAME, "Could not stat state %s", pstate->state_path);
+    return false;
+  }
+
+  if (!(fs = fopen(pstate->state_path, "r"))) {
+    TSDebug(PLUGIN_NAME, "Could not open state %s for reading", pstate->state_path);
+    return false;
+  }
+
+  now = time(NULL);
+
+  config_re = pcre_compile("^([^#].+?)\\s+(\\d+)\\s+(\\d+)\\s+(\\w+)\\s*$", 0, &errptr, &erroffset, NULL);
+  TSAssert(NULL != config_re);
+
+  while (fgets(line, LINE_MAX, fs) != NULL) {
+    TSDebug(PLUGIN_NAME, "state: processing: %d %s", ln, line);
+    ++ln;
+    rc = pcre_exec(config_re, NULL, line, strlen(line), 0, 0, ovector, OVECTOR_SIZE);
+
+    if (5 == rc) {
+      invalidate_t *const inv = (invalidate_t *)TSmalloc(sizeof(invalidate_t));
+      init_invalidate_t(inv);
+
+      pcre_get_substring(line, ovector, rc, 1, &(inv->regex_text));
+      inv->epoch  = atoi(line + ovector[4]);
+      inv->expiry = atoi(line + ovector[6]);
+
+      if (inv->expiry < now) {
+        TSDebug(PLUGIN_NAME, "state: skipping expired : '%s'", inv->regex_text);
+        free_invalidate_t(inv);
+        continue;
+      }
+
+      int const len          = ovector[9] - ovector[8];
+      char const *const type = line + ovector[8];
+
+      if (0 == strncasecmp(type, RESULT_STALE, len)) {
+        TSDebug(PLUGIN_NAME, "state: regex line set to result type %s: '%s'", RESULT_STALE, inv->regex_text);
+      } else if (0 == strncasecmp(type, RESULT_MISS, len)) {
+        TSDebug(PLUGIN_NAME, "state: regex line set to result type %s: '%s'", RESULT_MISS, inv->regex_text);
+        inv->new_result = TS_CACHE_LOOKUP_MISS;
+      } else {
+        TSDebug(PLUGIN_NAME, "state: unknown regex line result type '%.*s', skipping '%s'", len, type, inv->regex_text);
+      }
+
+      // iterate through the loaded config and try to merge
+      invalidate_t *iptr = *ilist;
+      while (1) {
+        if (0 == strcmp(inv->regex_text, iptr->regex_text)) {
+          if (iptr->expiry == inv->expiry && iptr->new_result == inv->new_result) {
+            TSDebug(PLUGIN_NAME, "state: restoring epoch for %s", iptr->regex_text);
+            iptr->epoch = inv->epoch;
+          }
+          break;
+        }
+
+        if (NULL == iptr->next) {
+          break;
+        }
+        iptr = iptr->next;
+      }
+
+      free_invalidate_t(inv);
```

Comment:
Why cant `inv` just be a stack variable?

Reply:
becaise free_invalidate_t free's the variable as well.  every single free_invalidate_t(i) would request TSfree(i) after it.